src/Codegen/Builders/ObjectBuilder.hack

```diff
@@ -0,0 +1,69 @@
+namespace Slack\GraphQL\Codegen;
+
+use namespace HH\Lib\Vec;
+use type Facebook\HackCodegen\{CodegenClass, CodegenMethod, HackBuilderValues, HackCodegenFactory};
+
+
+/**
+ * Builder for GraphQL objects and interfaces.
+ *
+ * The annotated Hack type should be either a class, interface, or shape.
+ */
+// TODO: It probably makes sense to have a separate builder for interfaces.
```

Comment:
Yep I think once we implement proper execution for interfaces (I'm working on it as part of #26 and #27) that will also affect what we need to codegen for interfaces -- the overlap between object types and interfaces should actually be pretty small, e.g. interface fields shouldn't really have any resolvers, they should delegate to object types at runtime.

Reply:
Oh interesting, I didn’t realize that. I was imagining it’d be the reverse: interface resolvers would contain common logic and objects would delegate to them if an interface field is being resolved. How does the interface resolver know which object resolver to delegate to? Would it need to introspect the Hack object?

Reply:
Yeah normally during execution, if we encounter a field whose return type is abstract (interface/union), then the first step when resolving the field is to narrow down its type to the correct subtype: https://spec.graphql.org/draft/#sel-JANRLFCJDDCAAEDBEB74H

This is needed for various things (e.g. for `__typename` to return the correct type name if it's included in the query; for `fragment Foo on TypeName` to correctly determine whether the fragment should be included or not...) but one consequence is that it is never needed to call any resolvers on an abstract type, since by the type we call any field resolver, we would have already narrowed it down to the correct non-abstract subtype.

Reply:
Makes sense. It doesn't sound too difficult to implement a naive version of that spec.

I'm thinking about the case where the object type is a subclass of a Hack type which implements the interface. E.g., say I have the following object hierarchy:

```
<<GraphQL\Interface('Animal')>>
interface Animal {}

<<GraphQL\Object('Feline')>>
class Feline implements Animal {}

class Puma extends Animal {}
```

So in this case, say the field resolver for `Animal` receives an instance of `Puma`. We should be able to determine that `Puma` is an instance of `Feline` and so call the generated resolver for `Feline`, per the spec (there is no resolver for `Puma`). So what this means is: we need an algorithm which walks the object hierarchy of Puma, looking for the "closest" parent class which has a field resolver (i.e., is annotated with a GQL attribute). This sounds a lot like "[Method Resolution Order](https://www.geeksforgeeks.org/method-resolution-order-in-python-inheritance/)" in Python.

Anyway, we can probably do something more naive for a v1. Mostly thinking aloud here.

Reply:
We could make things a lot easier if we enforced some rules:

1. if `<<GraphQL\Object>>` is used on some class, there must not be `<<GraphQL\Object>>` on any of its subtypes or supertypes
2. ban `<<GraphQL\Object>>` on Hack interfaces (only classes)

1 is not strictly necessary but would make things easier

2 might actually be necessary, because if we have something like

```Hack
<<GraphQL\Object('A')>>
interface A { ... }

<<GraphQL\Object('B')>>
interface B { ... }

class C implements A, B { ... }
```

then if we have an instance of `C`, there is no way to decide if its GraphQL type should be A or B, unless we come up with some super-arbitrary rule like "first one in alphabetical order wins"

Reply:
Wouldn't (1) make it illegal to have both an interface and a concrete implementation of that interface in the same schema? Since if B implements A, then B cannot be annotated with `<<GraphQL\Object>>`. But this means you can never write a query for B instead of A. For example, I think this schema would be illegal: https://github.com/apollographql/apollo-tooling/blob/master/__fixtures__/starwars/schema.graphql

I agree that (2) is probably necessary. How do y'all do this at FB?

Reply:
Oh nvm, I see you're talking about `ObjectType`, not both `ObjectType` and `InterfaceType`, in (1). I guess that would help - we'd just need to find the single class of a class's superclasses which is annotated with `GraphQL\ObjectType`, and use that as the resolver.

For (2), another option would be to ensure the class only extends a single interface tagged with `GraphQL\InterfaceType` during codegen.

Reply:
> Wouldn't (1) make it illegal to have both an interface and a concrete implementation of that interface in the same schema?

The parent type would have to be a `<<GraphQL\Interface>>`, not `<<GraphQL\Object>>`.

> How do y'all do this at FB?

I think we ban all the ambiguous cases (so probably the 2 rules I mentioned apply).

> For (2), another option would be to ensure the class only extends a single interface tagged with GraphQL\InterfaceType during codegen.

In my example the class `C` is not annotated with any GraphQL attribute. Would the GraphQL codegen enforce the rule for every class in the codebase, whether or not it has a GraphQL attribute on it?

Reply:
We'd have to enforce it for any class implementing an interface tagged with a GQL attribute, yeah. Which sounds untenable for a large codebase.